packages/calling-stateful-client/src/CallAgentDeclarativeCommon.ts

```diff
@@ -0,0 +1,287 @@
+        }
+      }
+    }
+
+    for (const externalCallsUpdatedListener of this._externalCallsUpdatedListeners) {
+      externalCallsUpdatedListener({ added: addedStatefulCall, removed: removedStatefulCall });
+    }
+  };
+
+  private setIncomingCallEnded = (incomingCallId: string, callEndReason: CallEndReason): void => {
+    const incomingCallSubscriber = this._incomingCallSubscribers.get(incomingCallId);
+    if (incomingCallSubscriber) {
+      incomingCallSubscriber.unsubscribe();
+      this._incomingCallSubscribers.delete(incomingCallId);
+    }
+    this._declarativeIncomingCalls.delete(incomingCallId);
+    this._context.setIncomingCallEnded(incomingCallId, callEndReason);
+  };
+
+  protected incomingCall = ({
+    incomingCall
+  }: {
+    incomingCall: IncomingCall | /* @conditional-compile-remove(teams-call) */ TeamsIncomingCall;
+  }): void => {
+    // Make sure to not subscribe to the incoming call if we are already subscribed to it.
+    if (!this._incomingCallSubscribers.has(incomingCall.id)) {
+      this._incomingCallSubscribers.set(
+        incomingCall.id,
+        new IncomingCallSubscriber(incomingCall, this.setIncomingCallEnded)
+      );
+    }
+    this._declarativeIncomingCalls.set(incomingCall.id, incomingCallDeclaratify(incomingCall, this._context));
+    this._context.setIncomingCall(convertSdkIncomingCallToDeclarativeIncomingCall(incomingCall));
+  };
+
+  protected addCall = (call: Call | /* @conditional-compile-remove(teams-call) */ TeamsCall): DeclarativeCallCommon => {
+    this._callSubscribers.get(call)?.unsubscribe();
+
+    // For API extentions we need to have the call in the state when we are subscribing as we may want to update the
+    // state during the subscription process in the subscriber so we add the call to state before subscribing.
+    this._context.setCall(convertSdkCallToDeclarativeCall(call));
+    this._callSubscribers.set(call, new CallSubscriber(call, this._context, this._internalContext));
+    return this.getOrCreateDeclarativeCall(call);
+  };
+
+  private getOrCreateDeclarativeCall = (
+    call: Call | /* @conditional-compile-remove(teams-call) */ TeamsCall
+  ): DeclarativeCallCommon => {
+    const declarativeCall = this._declarativeCalls.get(call);
+    if (declarativeCall) {
+      return declarativeCall;
+    }
+
+    /* @conditional-compile-remove(teams-call) */
+    if (!isACSCall(call)) {
+      const newDeclarativeCall = teamsCallDeclaratify(call, this._context);
+      this._declarativeCalls.set(call, newDeclarativeCall as DeclarativeTeamsCall);
+      return newDeclarativeCall;
+    }
+
+    const newDeclarativeCall = callDeclaratify(call, this._context);
+    this._declarativeCalls.set(call, newDeclarativeCall as DeclarativeCall);
+    return newDeclarativeCall;
+  };
+
+  public getCommon<
+    AgentType extends CallAgent | /* @conditional-compile-remove(teams-call) */ TeamsCallAgent,
+    P extends keyof CallAgent
```

Comment:
Two reasons for referring to `CallAgent` and `TeamsCallAgent` here (as opposed to `CallAgentCommon`).

1. Generic implementation for both. You just need type constraints.
2. Is actual behavior difference (you call `isACS*`).

For 1, let's think if we can extract this to a separate file.
For 2, we need to use protected methods.

Reply:
Tried some ways to resolve this...

The problem here is we don't have startCall/join function in CallAgentCommon, but they do exist in CallAgent | TeamsCallAgent

I will leave it for this PR, we can do some iterations later once we find a better way to refactor this

Reply:
If you look at CallDeclarativeCommon  - things are a lot better, that is because we can almost find every function there in callCommon. For CallAgent.. I think we can't do much with it

Reply:
From what I'm reading a callAgent is always a CallAgent or TeamsCallAgent, why do we use CallAgentCommon and not `CallAgent | TeamsCallAgent` (that would contain the correct overlaps)?

Reply:
This might need a separate comment thread but I want to confirm we _want_ a common abstract class. We're seemingly just using it to dedupe code? But going by the fact we have runtime checks to access properties the CallAgentCommon does not have would suggest there isn't enough overlap in these two classes to define them as two types of the same thing and warrant a base class.
Instead when common code exists between the two classes we could have pure helper functions.

I'm trying to find some good online resources on one vs the other but don't have too much time to invest in it and not finding much of use - but I'd love to hear your guys' opinions on this. (perhaps + @DominikMe)

Reply:
This is the long standing questions for every engineers - inheritance or composition. 

For inheritance, it adds more constrains and limits your logic all aligned to base class

For composition, it gives you more flexibility to grab those functions and use those in some other classes

The good news this is just the implementation details, for current scale of implementation, I would say both are good choice, composition implementation requires lots of big if else to make things happens, and that's why I chose more constrains from abstract class so I don't miss any part for CTE. 

if we are going to have 3 or 4+ classes or wanna reuse these startCall/join logics inside another class (just like callContext), we can always come back and refactor them into composition again